File: packages/search-lambda/src/duckdb_client.cpp
#include "duckdb_client.hpp"
#include <iostream>
#include <chrono>
#include <algorithm>
#include <unordered_set>
#include <unordered_map>
#include <cstdlib>
#include <sstream>

namespace fdnix {

    DuckDBClient::DuckDBClient(const std::string& db_path) 
        : db_path_(db_path), database_(nullptr), connection_(nullptr) {
        
        // Check if embeddings are enabled via environment variable
        const char* enable_embeddings = std::getenv("ENABLE_EMBEDDINGS");
        embeddings_enabled_ = enable_embeddings && 
                             (std::string(enable_embeddings) == "1" || 
                              std::string(enable_embeddings) == "true" || 
                              std::string(enable_embeddings) == "yes");
        
        std::cout << "DuckDBClient created for database: " << db_path_ 
                  << ", embeddings enabled: " << (embeddings_enabled_ ? "true" : "false") << std::endl;
    }

    DuckDBClient::~DuckDBClient() {
        // connection_ will be automatically cleaned up
        std::cout << "DuckDBClient destroyed" << std::endl;
    }

    bool DuckDBClient::initialize() {
        try {
            // Create DuckDB database instance with read-only config
            duckdb::DBConfig config;
            config.options.access_mode = duckdb::AccessMode::READ_ONLY;
            database_ = std::make_unique<duckdb::DuckDB>(db_path_, &config);
            connection_ = std::make_unique<duckdb::Connection>(*database_);
            
            // Load required extensions
            try {
                connection_->Query("LOAD fts;");
                std::cout << "FTS extension loaded successfully" << std::endl;
            } catch (const std::exception& e) {
                std::cerr << "Warning: Could not load FTS extension: " << e.what() << std::endl;
            }
            
            if (embeddings_enabled_) {
                try {
                    connection_->Query("LOAD vss;");
                    std::cout << "VSS extension loaded successfully" << std::endl;
                } catch (const std::exception& e) {
                    std::cerr << "Warning: Could not load VSS extension: " << e.what() << std::endl;
                    embeddings_enabled_ = false;  // Disable embeddings if VSS not available
                }
            }
            
            // Check if database has required tables
            auto result = connection_->Query("SELECT name FROM sqlite_master WHERE type='table' AND name='packages';");
            if (result->HasError() || result->RowCount() == 0) {
                std::cerr << "Required 'packages' table not found in database" << std::endl;
                return false;
            }
            
            // Check embeddings availability if enabled
            if (embeddings_enabled_) {
                embeddings_enabled_ = check_embeddings_availability();
                if (!embeddings_enabled_) {
                    std::cout << "Embeddings table not found or empty, falling back to FTS-only mode" << std::endl;
                }
            }
            
            std::cout << "DuckDB client initialized successfully (embeddings: " 
                      << (embeddings_enabled_ ? "enabled" : "disabled") << ")" << std::endl;
            return true;
            
        } catch (const std::exception& e) {
            std::cerr << "Error initializing DuckDB client: " << e.what() << std::endl;
            return false;
        }
    }

    SearchResults DuckDBClient::hybrid_search(const SearchParams& params, 
                                             const std::vector<double>& query_embedding) {
        auto start_time = std::chrono::high_resolution_clock::now();
        
        SearchResults results;
        
        try {
            if (embeddings_enabled_ && !query_embedding.empty()) {
                // Hybrid search mode
                results.search_type = "hybrid";
                
                // Perform vector search
                auto vector_results = vector_search(query_embedding, params.limit * 2);
                
                // Perform FTS search  
                auto fts_results = fts_search(params.query, params.limit * 2);
                
                // Combine using Reciprocal Rank Fusion
                results.packages = reciprocal_rank_fusion(
                    vector_results.packages, 
                    fts_results.packages
                );
            } else {
                // FTS-only search mode
                results.search_type = "fts";
                auto fts_results = fts_search(params.query, params.limit * 2);
                results.packages = fts_results.packages;
            }
            
            // Apply filters
            if (params.license_filter.has_value() || params.category_filter.has_value()) {
                results.packages.erase(
                    std::remove_if(results.packages.begin(), results.packages.end(),
                        [&params](const Package& pkg) {
                            if (params.license_filter.has_value() && 
                                pkg.license.find(params.license_filter.value()) == std::string::npos) {
                                return true;
                            }
                            // Category filtering would need additional metadata
                            return false;
                        }
                    ),
                    results.packages.end()
                );
            }
            
            // Apply offset and limit
            if (params.offset > 0 && params.offset < static_cast<int>(results.packages.size())) {
                results.packages.erase(results.packages.begin(), 
                                     results.packages.begin() + params.offset);
            }
            
            if (params.limit > 0 && static_cast<int>(results.packages.size()) > params.limit) {
                results.packages.resize(params.limit);
            }
            
            results.total_count = static_cast<int>(results.packages.size());
            
        } catch (const std::exception& e) {
            std::cerr << "Error in hybrid search: " << e.what() << std::endl;
            results.search_type = "error";
        }
        
        auto end_time = std::chrono::high_resolution_clock::now();
        auto duration = std::chrono::duration_cast<std::chrono::microseconds>(end_time - start_time);
        results.query_time_ms = duration.count() / 1000.0;
        
        return results;
    }

    SearchResults DuckDBClient::vector_search(const std::vector<double>& query_embedding, 
                                            int limit) {
        SearchResults results;
        results.search_type = "vector";
        
        if (!connection_ || !embeddings_enabled_ || query_embedding.empty()) {
            return results;
        }
        
        try {
            // Convert vector to DuckDB FLOAT array format
            std::ostringstream vec_str;
            vec_str << "[";
            for (size_t i = 0; i < query_embedding.size(); ++i) {
                if (i > 0) vec_str << ",";
                vec_str << query_embedding[i];
            }
            vec_str << "]";
            
            // Construct VSS query using DuckDB's vss_search function
            std::ostringstream query;
            query << "SELECT p.package_id, p.packageName, p.version, p.description, "
                  << "p.homepage, p.license, p.attributePath, d.distance "
                  << "FROM vss_search('embeddings_vss_idx', " << vec_str.str() << "::FLOAT[]) AS d "
                  << "JOIN embeddings e ON e.rowid = d.rowid "
                  << "JOIN packages p ON p.package_id = e.package_id "
                  << "ORDER BY d.distance ASC "
                  << "LIMIT " << limit << ";";
            
            auto result = connection_->Query(query.str());
            
            if (!result->HasError()) {
                for (size_t row = 0; row < result->RowCount(); ++row) {
                    Package pkg;
                    pkg.packageId = result->GetValue(0, row).ToString();
                    pkg.packageName = result->GetValue(1, row).ToString();
                    pkg.version = result->GetValue(2, row).ToString();
                    pkg.description = result->GetValue(3, row).ToString();
                    pkg.homepage = result->GetValue(4, row).ToString();
                    pkg.license = result->GetValue(5, row).ToString();
                    pkg.attributePath = result->GetValue(6, row).ToString();
                    
                    // Convert distance to similarity score (lower distance = higher score)
                    double distance = result->GetValue(7, row).GetValue<double>();
                    pkg.relevanceScore = 1.0 / (1.0 + distance);
                    
                    results.packages.push_back(pkg);
                }
            } else {
                std::cerr << "Vector search query failed: " << result->GetError() << std::endl;
            }
            
        } catch (const std::exception& e) {
            std::cerr << "Error in vector search: " << e.what() << std::endl;
        }
        
        results.total_count = static_cast<int>(results.packages.size());
        return results;
    }

    SearchResults DuckDBClient::fts_search(const std::string& query, int limit) {
        SearchResults results;
        results.search_type = "fts";
        
        if (!connection_ || query.empty()) {
            return results;
        }
        
        try {
            // Escape single quotes in the query for SQL safety
            std::string escaped_query = query;
            size_t pos = 0;
            while ((pos = escaped_query.find("'", pos)) != std::string::npos) {
                escaped_query.replace(pos, 1, "''");
                pos += 2;
            }
            
            // Construct FTS query using DuckDB FTS with BM25 scoring
            std::ostringstream sql_query;
            sql_query << "SELECT p.package_id, p.packageName, p.version, p.description, "
                      << "p.homepage, p.license, p.attributePath, fts.score "
                      << "FROM (SELECT package_id, fts_main_packages_fts_source.match_bm25(package_id, '" << escaped_query << "') AS score FROM packages_fts_source) fts "
                      << "JOIN packages p ON p.package_id = fts.package_id "
                      << "WHERE fts.score IS NOT NULL "
                      << "ORDER BY fts.score DESC "
                      << "LIMIT " << limit << ";";
            
            auto result = connection_->Query(sql_query.str());
            
            if (!result->HasError()) {
                for (size_t row = 0; row < result->RowCount(); ++row) {
                    Package pkg;
                    pkg.packageId = result->GetValue(0, row).ToString();
                    pkg.packageName = result->GetValue(1, row).ToString();
                    pkg.version = result->GetValue(2, row).ToString();
                    pkg.description = result->GetValue(3, row).ToString();
                    pkg.homepage = result->GetValue(4, row).ToString();
                    pkg.license = result->GetValue(5, row).ToString();
                    pkg.attributePath = result->GetValue(6, row).ToString();
                    
                    // Use BM25 score directly as relevance score
                    pkg.relevanceScore = result->GetValue(7, row).GetValue<double>();
                    
                    results.packages.push_back(pkg);
                }
            } else {
                std::cerr << "FTS search query failed: " << result->GetError() << std::endl;
                
                // Fallback to simple LIKE search if FTS fails
                std::ostringstream fallback_query;
                fallback_query << "SELECT package_id, packageName, version, description, "
                              << "homepage, license, attributePath, 1.0 as score "
                              << "FROM packages "
                              << "WHERE packageName ILIKE '%" << escaped_query << "%' "
                              << "OR description ILIKE '%" << escaped_query << "%' "
                              << "ORDER BY CASE WHEN packageName ILIKE '%" << escaped_query << "%' THEN 1 ELSE 2 END, "
                              << "packageName "
                              << "LIMIT " << limit << ";";
                
                auto fallback_result = connection_->Query(fallback_query.str());
                if (!fallback_result->HasError()) {
                    for (size_t row = 0; row < fallback_result->RowCount(); ++row) {
                        Package pkg;
                        pkg.packageId = fallback_result->GetValue(0, row).ToString();
                        pkg.packageName = fallback_result->GetValue(1, row).ToString();
                        pkg.version = fallback_result->GetValue(2, row).ToString();
                        pkg.description = fallback_result->GetValue(3, row).ToString();
                        pkg.homepage = fallback_result->GetValue(4, row).ToString();
                        pkg.license = fallback_result->GetValue(5, row).ToString();
                        pkg.attributePath = fallback_result->GetValue(6, row).ToString();
                        pkg.relevanceScore = 1.0 - (row * 0.1);  // Simple decreasing score
                        
                        results.packages.push_back(pkg);
                    }
                }
            }
            
        } catch (const std::exception& e) {
            std::cerr << "Error in FTS search: " << e.what() << std::endl;
        }
        
        results.total_count = static_cast<int>(results.packages.size());
        return results;
    }

    bool DuckDBClient::health_check() {
        if (!connection_) {
            return false;
        }
        
        try {
            // Execute a simple query to verify database health
            auto result = connection_->Query("SELECT 1;");
            return !result->HasError();
        } catch (const std::exception& e) {
            std::cerr << "Health check failed: " << e.what() << std::endl;
            return false;
        }
    }

    std::vector<Package> DuckDBClient::reciprocal_rank_fusion(
        const std::vector<Package>& vector_results,
        const std::vector<Package>& fts_results,
        double k) {
        
        std::unordered_map<std::string, Package> combined_packages;
        std::unordered_map<std::string, double> rrf_scores;
        
        // Process vector results
        for (size_t i = 0; i < vector_results.size(); ++i) {
            const auto& pkg = vector_results[i];
            const std::string key = pkg.packageId.empty() ? pkg.packageName : pkg.packageId;
            
            // RRF score: 1 / (k + rank)
            double score = 1.0 / (k + i + 1);
            
            combined_packages[key] = pkg;
            rrf_scores[key] = score;
        }
        
        // Process FTS results and add/merge scores
        for (size_t i = 0; i < fts_results.size(); ++i) {
            const auto& pkg = fts_results[i];
            const std::string key = pkg.packageId.empty() ? pkg.packageName : pkg.packageId;
            
            // RRF score: 1 / (k + rank)
            double score = 1.0 / (k + i + 1);
            
            if (combined_packages.find(key) != combined_packages.end()) {
                // Package exists, add to RRF score
                rrf_scores[key] += score;
            } else {
                // New package from FTS
                combined_packages[key] = pkg;
                rrf_scores[key] = score;
            }
        }
        
        // Convert to vector and sort by RRF score
        std::vector<Package> result;
        for (auto& [key, pkg] : combined_packages) {
            pkg.relevanceScore = rrf_scores[key];
            result.push_back(pkg);
        }
        
        // Sort by RRF score (descending)
        std::sort(result.begin(), result.end(),
            [](const Package& a, const Package& b) {
                return a.relevanceScore > b.relevanceScore;
            });
        
        return result;
    }

    std::vector<Package> DuckDBClient::combine_and_rank_results(
        const std::vector<Package>& vector_results,
        const std::vector<Package>& fts_results,
        double /*vector_weight*/,
        double /*fts_weight*/) {
        
        // Use RRF for combining results
        return reciprocal_rank_fusion(vector_results, fts_results);
    }

    bool DuckDBClient::check_embeddings_availability() {
        if (!connection_) {
            return false;
        }
        
        try {
            // Check if embeddings table exists
            auto result = connection_->Query("SELECT name FROM sqlite_master WHERE type='table' AND name='embeddings';");
            if (result->HasError() || result->RowCount() == 0) {
                return false;
            }
            
            // Check if embeddings table has data
            auto count_result = connection_->Query("SELECT COUNT(*) FROM embeddings WHERE vector IS NOT NULL;");
            if (count_result->HasError()) {
                return false;
            }
            
            int64_t count = count_result->GetValue(0, 0).GetValue<int64_t>();
            return count > 0;
            
        } catch (const std::exception& e) {
            std::cerr << "Error checking embeddings availability: " << e.what() << std::endl;
            return false;
        }
    }

} // namespace fdnix

File: packages/search-lambda/src/duckdb_client.rs
use duckdb::{Connection, Result as DuckDBResult};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::env;
use std::time::Instant;
use thiserror::Error;
use tracing::{info, warn, error};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Package {
    pub package_id: String,
    pub package_name: String,
    pub version: String,
    pub description: String,
    pub homepage: String,
    pub license: String,
    pub attribute_path: String,
    pub relevance_score: f64,
}

#[derive(Debug, Clone)]
pub struct SearchParams {
    pub query: String,
    pub limit: i32,
    pub offset: i32,
    pub license_filter: Option<String>,
    pub category_filter: Option<String>,
}

#[derive(Debug, Clone)]
pub struct SearchResults {
    pub packages: Vec<Package>,
    pub total_count: i32,
    pub query_time_ms: f64,
    pub search_type: String,
}

#[derive(Error, Debug)]
pub enum DuckDBClientError {
    #[error("Database error: {0}")]
    DatabaseError(#[from] duckdb::Error),
    #[error("Connection not initialized")]
    NotInitialized,
    #[error("Required table not found: {0}")]
    TableNotFound(String),
    #[error("Query failed: {0}")]
    QueryFailed(String),
}

pub struct DuckDBClient {
    db_path: String,
    connection: Option<Connection>,
    embeddings_enabled: bool,
}

impl DuckDBClient {
    pub fn new(db_path: &str) -> Result<Self, DuckDBClientError> {
        let embeddings_enabled = env::var("ENABLE_EMBEDDINGS")
            .map(|val| val == "1" || val.to_lowercase() == "true" || val.to_lowercase() == "yes")
            .unwrap_or(false);

        info!(
            "DuckDBClient created for database: {}, embeddings enabled: {}",
            db_path, embeddings_enabled
        );

        Ok(DuckDBClient {
            db_path: db_path.to_string(),
            connection: None,
            embeddings_enabled,
        })
    }

    pub fn initialize(&mut self) -> Result<bool, DuckDBClientError> {
        // Open database connection with read-only access
        let conn = Connection::open_with_flags(
            &self.db_path,
            duckdb::Config::default()
                .access_mode(duckdb::AccessMode::ReadOnly)?
        )?;

        // Set home directory to prevent extension loading issues
        if let Err(e) = conn.execute("SET home_directory = '/tmp';", []) {
            warn!("Could not set home directory: {}", e);
        }

        info!("Using DuckDB with built-in extensions: FTS, JSON");

        // Check if database has required tables
        let mut stmt = conn.prepare("SELECT name FROM sqlite_master WHERE type='table' AND name='packages';")?;
        let rows: Result<Vec<String>, _> = stmt
            .query_map([], |row| Ok(row.get::<_, String>(0)?))?
            .collect();

        match rows {
            Ok(names) if names.is_empty() => {
                error!("Required 'packages' table not found in database");
                return Err(DuckDBClientError::TableNotFound("packages".to_string()));
            }
            Ok(_) => {
                info!("Found required 'packages' table");
            }
            Err(e) => {
                error!("Error checking for packages table: {}", e);
                return Err(DuckDBClientError::DatabaseError(e));
            }
        }

        // Check embeddings availability if enabled
        if self.embeddings_enabled {
            if self.check_embeddings_availability(&conn) {
                info!("Embeddings available and enabled");
            } else {
                warn!("VSS extension not available in this build - disabling embeddings, using FTS-only mode");
                self.embeddings_enabled = false;
            }
        }

        self.connection = Some(conn);

        info!(
            "DuckDB client initialized successfully (embeddings: {})",
            if self.embeddings_enabled { "enabled" } else { "disabled" }
        );

        Ok(true)
    }

    pub fn hybrid_search(
        &self,
        params: &SearchParams,
        query_embedding: Option<&[f64]>,
    ) -> Result<SearchResults, DuckDBClientError> {
        let start_time = Instant::now();
        let mut results = SearchResults {
            packages: Vec::new(),
            total_count: 0,
            query_time_ms: 0.0,
            search_type: "unknown".to_string(),
        };

        let _conn = self.connection.as_ref()
            .ok_or(DuckDBClientError::NotInitialized)?;

        if self.embeddings_enabled && query_embedding.is_some() && !query_embedding.unwrap().is_empty() {
            // Hybrid search mode
            results.search_type = "hybrid".to_string();

            // Perform vector search
            let vector_results = self.vector_search(query_embedding.unwrap(), params.limit * 2)?;

            // Perform FTS search
            let fts_results = self.fts_search(&params.query, params.limit * 2)?;

            // Combine using Reciprocal Rank Fusion
            results.packages = self.reciprocal_rank_fusion(&vector_results.packages, &fts_results.packages, 60.0);
        } else {
            // FTS-only search mode
            results.search_type = "fts".to_string();
            let fts_results = self.fts_search(&params.query, params.limit * 2)?;
            results.packages = fts_results.packages;
        }

        // Apply filters
        if params.license_filter.is_some() || params.category_filter.is_some() {
            results.packages.retain(|pkg| {
                if let Some(license_filter) = &params.license_filter {
                    if !pkg.license.contains(license_filter) {
                        return false;
                    }
                }
                // Category filtering would need additional metadata
                true
            });
        }

        // Apply offset and limit
        if params.offset > 0 && (params.offset as usize) < results.packages.len() {
            results.packages.drain(0..(params.offset as usize));
        }

        if params.limit > 0 && results.packages.len() > (params.limit as usize) {
            results.packages.truncate(params.limit as usize);
        }

        results.total_count = results.packages.len() as i32;
        results.query_time_ms = start_time.elapsed().as_millis() as f64;

        Ok(results)
    }

    pub fn vector_search(&self, query_embedding: &[f64], limit: i32) -> Result<SearchResults, DuckDBClientError> {
        let mut results = SearchResults {
            packages: Vec::new(),
            total_count: 0,
            query_time_ms: 0.0,
            search_type: "vector".to_string(),
        };

        let conn = self.connection.as_ref()
            .ok_or(DuckDBClientError::NotInitialized)?;

        if !self.embeddings_enabled || query_embedding.is_empty() {
            return Ok(results);
        }

        // Convert vector to DuckDB FLOAT array format
        let vec_str = format!(
            "[{}]",
            query_embedding
                .iter()
                .map(|v| v.to_string())
                .collect::<Vec<_>>()
                .join(",")
        );

        // Construct VSS query using DuckDB's vss_search function
        let query = format!(
            r#"
            SELECT p.package_id, p.packageName, p.version, p.description,
                   p.homepage, p.license, p.attributePath, d.distance
            FROM vss_search('embeddings_vss_idx', {}::FLOAT[]) AS d
            JOIN embeddings e ON e.rowid = d.rowid
            JOIN packages p ON p.package_id = e.package_id
            ORDER BY d.distance ASC
            LIMIT {};
            "#,
            vec_str, limit
        );

        let mut stmt = conn.prepare(&query)?;
        let rows = stmt.query_map([], |row| {
            let distance: f64 = row.get(7)?;
            Ok(Package {
                package_id: row.get(0)?,
                package_name: row.get(1)?,
                version: row.get(2)?,
                description: row.get(3)?,
                homepage: row.get(4)?,
                license: row.get(5)?,
                attribute_path: row.get(6)?,
                // Convert distance to similarity score (lower distance = higher score)
                relevance_score: 1.0 / (1.0 + distance),
            })
        })?;

        results.packages = rows.collect::<Result<Vec<_>, _>>()?;
        results.total_count = results.packages.len() as i32;

        Ok(results)
    }

    pub fn fts_search(&self, query: &str, limit: i32) -> Result<SearchResults, DuckDBClientError> {
        let mut results = SearchResults {
            packages: Vec::new(),
            total_count: 0,
            query_time_ms: 0.0,
            search_type: "fts".to_string(),
        };

        let conn = self.connection.as_ref()
            .ok_or(DuckDBClientError::NotInitialized)?;

        if query.is_empty() {
            return Ok(results);
        }

        // Escape single quotes in the query for SQL safety
        let escaped_query = query.replace("'", "''");

        // Construct FTS query using DuckDB FTS with BM25 scoring
        let fts_query = format!(
            r#"
            SELECT p.package_id, p.packageName, p.version, p.description,
                   p.homepage, p.license, p.attributePath, fts.score
            FROM (SELECT package_id, fts_main_packages_fts_source.match_bm25(package_id, '{}') AS score 
                  FROM packages_fts_source) fts
            JOIN packages p ON p.package_id = fts.package_id
            WHERE fts.score IS NOT NULL
            ORDER BY fts.score DESC
            LIMIT {};
            "#,
            escaped_query, limit
        );

        let mut stmt = match conn.prepare(&fts_query) {
            Ok(stmt) => stmt,
            Err(e) => {
                error!("FTS search query failed: {}, falling back to simple search", e);
                return self.fallback_search(&escaped_query, limit);
            }
        };

        let rows = stmt.query_map([], |row| {
            Ok(Package {
                package_id: row.get(0)?,
                package_name: row.get(1)?,
                version: row.get(2)?,
                description: row.get(3)?,
                homepage: row.get(4)?,
                license: row.get(5)?,
                attribute_path: row.get(6)?,
                relevance_score: row.get(7)?, // BM25 score
            })
        })?;

        results.packages = rows.collect::<Result<Vec<_>, _>>()?;
        results.total_count = results.packages.len() as i32;

        Ok(results)
    }

    fn fallback_search(&self, escaped_query: &str, limit: i32) -> Result<SearchResults, DuckDBClientError> {
        let mut results = SearchResults {
            packages: Vec::new(),
            total_count: 0,
            query_time_ms: 0.0,
            search_type: "fts".to_string(),
        };

        let conn = self.connection.as_ref()
            .ok_or(DuckDBClientError::NotInitialized)?;

        // Fallback to simple LIKE search if FTS fails
        let fallback_query = format!(
            r#"
            SELECT package_id, packageName, version, description,
                   homepage, license, attributePath, 1.0 as score
            FROM packages
            WHERE packageName ILIKE '%{}%'
               OR description ILIKE '%{}%'
            ORDER BY CASE WHEN packageName ILIKE '%{}%' THEN 1 ELSE 2 END,
                     packageName
            LIMIT {};
            "#,
            escaped_query, escaped_query, escaped_query, limit
        );

        let mut stmt = conn.prepare(&fallback_query)?;
        let rows = stmt.query_map([], |row| {
            Ok(Package {
                package_id: row.get(0)?,
                package_name: row.get(1)?,
                version: row.get(2)?,
                description: row.get(3)?,
                homepage: row.get(4)?,
                license: row.get(5)?,
                attribute_path: row.get(6)?,
                relevance_score: row.get(7)?,
            })
        })?;

        let mut packages: Vec<Package> = rows.collect::<Result<Vec<_>, _>>()?;
        
        // Assign decreasing scores for fallback results
        for (i, pkg) in packages.iter_mut().enumerate() {
            pkg.relevance_score = 1.0 - (i as f64 * 0.1);
        }

        results.packages = packages;
        results.total_count = results.packages.len() as i32;

        Ok(results)
    }

    pub fn health_check(&self) -> bool {
        if let Some(conn) = &self.connection {
            match conn.execute("SELECT 1;", []) {
                Ok(_) => true,
                Err(e) => {
                    error!("Health check failed: {}", e);
                    false
                }
            }
        } else {
            false
        }
    }

    fn reciprocal_rank_fusion(
        &self,
        vector_results: &[Package],
        fts_results: &[Package],
        k: f64,
    ) -> Vec<Package> {
        let mut combined_packages: HashMap<String, Package> = HashMap::new();
        let mut rrf_scores: HashMap<String, f64> = HashMap::new();

        // Process vector results
        for (i, pkg) in vector_results.iter().enumerate() {
            let key = if pkg.package_id.is_empty() {
                &pkg.package_name
            } else {
                &pkg.package_id
            };

            // RRF score: 1 / (k + rank)
            let score = 1.0 / (k + i as f64 + 1.0);

            combined_packages.insert(key.clone(), pkg.clone());
            rrf_scores.insert(key.clone(), score);
        }

        // Process FTS results and add/merge scores
        for (i, pkg) in fts_results.iter().enumerate() {
            let key = if pkg.package_id.is_empty() {
                &pkg.package_name
            } else {
                &pkg.package_id
            };

            // RRF score: 1 / (k + rank)
            let score = 1.0 / (k + i as f64 + 1.0);

            if let Some(existing_score) = rrf_scores.get_mut(key) {
                // Package exists, add to RRF score
                *existing_score += score;
            } else {
                // New package from FTS
                combined_packages.insert(key.clone(), pkg.clone());
                rrf_scores.insert(key.clone(), score);
            }
        }

        // Convert to vector and sort by RRF score
        let mut result: Vec<Package> = combined_packages
            .into_iter()
            .map(|(key, mut pkg)| {
                pkg.relevance_score = rrf_scores[&key];
                pkg
            })
            .collect();

        // Sort by RRF score (descending)
        result.sort_by(|a, b| b.relevance_score.partial_cmp(&a.relevance_score).unwrap());

        result
    }

    fn check_embeddings_availability(&self, conn: &Connection) -> bool {
        // Check if embeddings table exists
        let table_check = conn.prepare("SELECT name FROM sqlite_master WHERE type='table' AND name='embeddings';");
        if let Ok(mut stmt) = table_check {
            let rows: Result<Vec<String>, _> = stmt
                .query_map([], |row| Ok(row.get::<_, String>(0)?))
                .unwrap()
                .collect();

            match rows {
                Ok(names) if names.is_empty() => return false,
                Err(_) => return false,
                _ => {}
            }
        } else {
            return false;
        }

        // Check if embeddings table has data
        let count_check = conn.prepare("SELECT COUNT(*) FROM embeddings WHERE vector IS NOT NULL;");
        if let Ok(mut stmt) = count_check {
            if let Ok(mut rows) = stmt.query([]) {
                if let Ok(Some(row)) = rows.next() {
                    if let Ok(count) = row.get::<_, i64>(0) {
                        return count > 0;
                    }
                }
            }
        }

        false
    }
}